plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,482 @@
+							&out_max_transfer_size,
+							NULL,
+							&ctx->error)) {
+		g_debug ("error: couldn't parse response message: %s\n", ctx->error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	total_sending_numbers = (g_bytes_get_size (ctx->blob)) / (out_max_transfer_size);
+	if ((g_bytes_get_size (ctx->blob)) % (out_max_transfer_size) != 0)
+		total_sending_numbers ++;
+
+	ctx->chunks = fu_chunk_array_new_from_bytes (ctx->blob,
+						     0x00,	/* start addr */
+						     0x00,	/* page_sz */
+						     out_max_transfer_size);
+	chk = g_ptr_array_index (ctx->chunks, 0);
+	request = mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *) fu_chunk_get_data (chk), NULL);
+	mbim_device_command (ctx->mbim_device,
+			     request,
+			     10,
+			     NULL,
+			     (GAsyncReadyCallback) fu_mbim_qdu_updater_file_write_ready,
+			     ctx);
+}
+
+static void
+fu_mbim_qdu_updater_session_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+	g_autoptr(MbimMessage) response = NULL;
+	g_autoptr(MbimMessage) request = NULL;
+
+	response = mbim_device_command_finish (device, res, &ctx->error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &ctx->error)) {
+		g_debug ("error: operation failed: %s\n", ctx->error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_update_session_response_parse (response,
+							     NULL,
+							     NULL,
+							     NULL,
+							     NULL,
+							     NULL,
+							     NULL,
+							     &ctx->error)) {
+		g_debug ("error: couldn't parse response message: %s\n", ctx->error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	g_debug ("[%s] Successfully request modem to update session", mbim_device_get_path_display (device));
+
+	request = mbim_message_qdu_file_open_set_new (MBIM_QDU_FILE_TYPE_LITTLE_ENDIAN_PACKAGE, g_bytes_get_size (ctx->blob), NULL);
+	mbim_device_command (device,
+			     request,
+			     10,
+			     NULL,
+			     (GAsyncReadyCallback) fu_mbim_qdu_updater_file_open_ready,
+			     ctx);
+}
+
+static void
+fu_mbim_qdu_updater_set_update_session (MbimDevice *device, WriteContext *ctx)
+{
+  	g_autoptr(MbimMessage) request = NULL;
+
+	request = mbim_message_qdu_update_session_set_new (MBIM_QDU_SESSION_ACTION_START, MBIM_QDU_SESSION_TYPE_LE, NULL);
+
+	mbim_device_command (device,
+			     request,
+			     10,
+			     NULL,
+			     (GAsyncReadyCallback) fu_mbim_qdu_updater_session_ready,
+			     ctx);
+}
+
+static GArray *
+fu_mbim_qdu_updater_get_checksum (GBytes *blob)
+{
+	gsize file_size;
+	gsize hash_size;
+	GArray *digest;
+	g_autoptr(GChecksum) checksum = NULL;
+
+	/* get checksum, to be used as unique id */
+	file_size = g_bytes_get_size (blob);
+	hash_size = g_checksum_type_get_length (G_CHECKSUM_SHA1);
+	checksum = g_checksum_new (G_CHECKSUM_SHA1);
```

Comment:
can you put a note here explaining why SHA1 and not SHA256 otherwise I'll get asked by the RH security team. Thanks!

Reply:
OK, in order to get faster computing speed, I will add the note, thanks.

Reply:
hmm, that's not much of reason. Can we use SHA256 at all? All SHA1 use needs to be approved by the security team.

Reply:
OK, I think SHA256 is also possible and I will modify it, thanks.